resolver/map.go

```diff
@@ -136,3 +136,134 @@ func (a *AddressMap) Values() []any {
+// comparison is o(n), where n is the number of unique addresses in the endpoint
+// rather than having to sort a list on each comparison which is n(log(n))
+// (cannot use slices for the latter case or maps for the former case as map
+// keys, so needs to be a list).
+type EndpointMap struct {
+	endpoints endpointNodeList // Cannot use maps as a map key. Thus, keep a list, and o(n) search. Not on RPC fast path so efficiency not as important.
+}
+
+// NewEndpointMap returns a new EndpointMap.
+func NewEndpointMap() *EndpointMap {
+	return &EndpointMap{}
+}
+
+// Get returns the value for the address in the map, if present.
+func (em *EndpointMap) Get(e Endpoint) (value any, ok bool) {
+	en := toEndpointNode(e)
+	if i := em.endpoints.find(en); i != -1 {
+		return em.endpoints[i].val, true
+	}
+	return nil, false
+}
+
+// Set updates or adds the value to the address in the map.
+func (em *EndpointMap) Set(e Endpoint, value any) {
+	en := toEndpointNode(e)
+	if i := em.endpoints.find(en); i != -1 {
+		em.endpoints[i].val = value
+		return
+	}
+	en.val = value
+	em.endpoints = append(em.endpoints, en)
+}
+
+// Len returns the number of entries in the map.
+func (em *EndpointMap) Len() int {
+	return len(em.endpoints)
+}
+
+// Keys returns a slice of all current map keys, as endpoints specifying the
+// addresses present in the endpoint keys, in which uniqueness is determined by
+// the unordered set of addresses. Thus, endpoint information returned is not
+// the full endpoint data but can be used for EndpointMap accesses.
+func (em *EndpointMap) Keys() []Endpoint { // TODO: Persist the whole endpoint data to return in nodes? No use case now, but one could come up in future.
+	ret := make([]Endpoint, 0, len(em.endpoints))
+	for _, en := range em.endpoints {
+		var endpoint Endpoint
+		for addr, count := range en.addrs {
+			for i := 0; i < count; i++ {
+				endpoint.Addresses = append(endpoint.Addresses, Address{
+					Addr: addr,
+				})
+			}
+		}
+		ret = append(ret, endpoint)
+	}
+	return ret
+}
+
+// Values returns a slice of all current map values.
+func (em *EndpointMap) Values() []any {
+	ret := make([]any, 0, len(em.endpoints))
+	for _, en := range em.endpoints {
+		ret = append(ret, en.val)
+	}
+	return ret
+}
+
+// find returns the index of endpoint in the endpointNodeList slice, or -1 if
+// not present. The comparisons are done on the unordered set of the addresses
+// within endpoint.
+func (l endpointNodeList) find(e endpointNode) int {
+	for i, endpoint := range l {
+		if endpoint.Equal(e) {
+			return i
+		}
+	}
+	return -1
+}
+
+// Delete removes the specified endpoint from the map.
+func (em *EndpointMap) Delete(e Endpoint) {
+	en := toEndpointNode(e)
+	entry := em.endpoints.find(en)
+	if entry == -1 {
+		return
+	}
+	if len(em.endpoints) == 1 {
+		em.endpoints = nil
+	}
+	copy(em.endpoints[entry:], em.endpoints[entry+1:])
+	em.endpoints = em.endpoints[:len(em.endpoints)-1]
```

Comment:
Another option would be to make `endpoints` a `map[*endpointNode]struct{}`.  `find` would return the `*endpointNode` if the endpoint can be found.  `Delete` can be a simple `delete(endpoints, entry)`.

Reply:
The issue I see with using a map of unordered addresses in a map itself is it's not allowed (similar to how you exclude attributes in your address map, since map[key] underlying using something similar to cmp.Equal, which panics on a map key, as you mentioned offline). Thus, I made it a list, with o(n) operations not on the RPC path, so I can do comparisons on each node. I don't think this would work with a map, unless you can overwrite the _, ok := map[key] operation, which I have not have found a way to do.

Reply:
Maybe you weren't understanding my suggestion.  Pointers are valid for map keys, and are compared using pointer equality:

https://go.dev/play/p/VWhAQUEueKR

We wouldn't do lookups using the key, though, even though a map was used.  But this simplifies deletions to: 1. find the entry to be deleted by scanning all the entries; 2. `delete(endpoints, entryPtr)`

Reply:
Ah, I see. Switched to that.